### ISSUE TITLE
More automatic compatibility with python 3

### ISSUE THREAD

**Issue by lamblin** [Theano/Theano#3900]
Since Travis does not test everything, we should make sure we use the python 3 behaviour when possible, by using imports from `__future__`. In particular, I think we should add to all (or most) files:
```python
from __future__ import absolute_import
from __future__ import division
from __future__ import print_function
```
And maybe `from __future__ import unicode_literals` as well.

**Comment by dwf** [comment]
Doesn't `division` alter behaviour in a way that Theano actually alters itself, depending if PY2 or PY3? (I could be wrong about this, but I thought that Theano mirrored the platform behaviour of integer division, or maybe that's numpy).

**Comment by lamblin** [comment]
Right, Theano seems to mimic the platform division for the `/` operator on tensor variables.
What I had in mind is, there are places in the tests, and for some shape inference computations, where we divide regular Python integers, and these would regularly raise deprecation warnings at least, potentially different behaviours for python 3.
Do the `__future__` imports affect only the current file? Or do they modify everything that gets imported afterwards?

**Comment by dwf** [comment]
They only affect the module they are specified in, AFAIK.

**Comment by lamblin** [comment]
So, in the worst case, we would only have to define the division for tensors in a separate file that would not import `division`, and we could still import it elsewhere.

**Comment by Sentient07** [comment]
@lamblin , can i take this up ?

**Comment by nouiz** [comment]
for the print, we already to this in many files. But maybe not all files.
(especially new files could have this missing)

@lamblin, I don't understand what you try to address. Theano defile
`__truediv__` and `__div__` Also, we do something "probably wrongly" in
tensor/var.py that try to set `__truediv__` to `__div__` for python 3.

**Comment by lamblin** [comment]
@nouiz For division, I'd like to address mainly the computation of shapes on integers, which happens in particular in tests, but possibly in `infer_shape` methods when called on Python integers. We may have to check again the behaviour on Theano tensors for python3, you're right.
@Sentient07 I think you can, unless @nouiz objects.

**Comment by nouiz** [comment]
I don't object, I'm just not sure about the division. The rest I'm pretty
sure it is good.

**Comment by lamblin** [comment]
OK, so some testing is in order to make sure the division does not mess things up. I think the existing tests for elemwise would be enough.
I actually think we should mandate those imports on new files through a script like https://pypi.python.org/pypi/flake8-future-import

**Comment by Sentient07** [comment]
So, from what i understand, we've to ensure that the ```__future__``` module is used whenever we're in need of classic division to be performed ?

**Comment by Sentient07** [comment]
@lamblin , that script checks if imports from __future__ are present or missing. Should I go according to that and make those imports wherever necessary ? I fear I might break code at places where some imports are not desired.

**Comment by lamblin** [comment]
I think that the unit tests will catch those issues, if adding those statements break code.
In fact, if things break because of these statements, that means that there is probably an issue with python 3 already.
So I would say go ahead and add the import statements, and let's see if it breaks something.

**Comment by Sentient07** [comment]
@lamblin sorry for the delayed reply. I had added the imports to all the files and few files had issues with division import. I had removed it from those files where the tests failed.

**Comment by tsirif** [comment]
Is this closed with 0.8.0?

**Comment by nouiz** [comment]
No, the PR isn't merged. This wasn't needed for the release any. I made a comment on the PR to help get it merged soon.
